src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ **/
+
+package io.confluent.copycat.hdfs;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.ipc.RemoteException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.copycat.errors.CopycatException;
+import org.apache.kafka.copycat.errors.TransientException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+
+import io.confluent.copycat.hdfs.WALFile.Reader;
+import io.confluent.copycat.hdfs.WALFile.Writer;
+
+public class WAL {
+
+  private static final Logger log = LoggerFactory.getLogger(WAL.class);
+  private static final String leaseException =
+      "org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException";
+  private static final long MAX_SLEEP_INERVAL_MILLIS = 16000L;
+  private static final int MAX_NUM_RETRY = 5;
+
+  private WALFile.Writer writer = null;
+  private WALFile.Reader reader = null;
+  private FileSystem fs = null;
+  private String logFile = null;
+  private Configuration conf = null;
+  private long sleepIntervalMillis = 1000L;
+  private int num_retry = 0;
+
+  public WAL(Configuration conf, String url, String topicsDir, TopicPartition topicPart)
+      throws CopycatException {
+    try {
+      this.conf = conf;
+      fs = FileSystem.newInstance(URI.create(url), conf);
+      logFile = FileUtils.logFileName(url, topicsDir, topicPart);
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws CopycatException {
+    try {
+      acquireLease();
+      WALEntry key = new WALEntry(tempFile);
+      WALEntry value = new WALEntry(committedFile);
+      writer.append(key, value);
+      writer.hsync();
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private void acquireLease() throws CopycatException {
+    long initialSleepIntervalMillis = sleepIntervalMillis;
+    while (sleepIntervalMillis < MAX_SLEEP_INERVAL_MILLIS) {
+      try {
+        if (writer == null) {
+          writer = WALFile.createWriter(conf,
+                                        Writer.file(new Path(logFile)),
+                                        Writer.appendIfExists(true));
+        }
+        log.info("Successfully acquired lease for {}", logFile);
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMillis = sleepIntervalMillis * 2;
```

Comment:
Since we multiply by 2 here and then immediately sleep, won't this exceed the `MAX_SLEEP_INTERVAL_MILLIS` at least once?